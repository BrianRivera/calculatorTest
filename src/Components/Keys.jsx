import React, { memo } from 'react'
//using memo for eviting the refresh of the component
export const Keys = memo(({setExercise}) => {

/*
used to decrease the identical code of the buttons, the return is traversed
the lists using the .map and creating the buttons
*/
    
const keyNumeric = [1,2,3,4,5,6,7,8,9,0,'.'];
const keySymbol = ['+','-','*','/'];

/*
function applied to all buttons with the exception of the AC and = to update the status of the exercise:
1.-take button value

2.-use the property of the setState (setExercise) that fetches the current value of the state 

3.-validates if the last digit of the exercise and the value of the button 
are mathematical symbols, if they are, replace the last digit with the new value

4.-valid if the current exercise is '0' and the new value is '.' or some symbol of mathematical operation to add the
new courage to exercise, if not '.' replace 0 with the new number

5.-copy the previous values ​​of the state and replace the value of the exercise, with the new value
*/

const handleSymbolNumber = (e) => {
    const value = e.target.value;

    setExercise(last => {
        
        (keySymbol.includes(last.ex.substr(-1)) && keySymbol.includes(value)) 
        && (last.ex = last.ex.slice(0, -1));

        const lastEx = (last.ex === '0' && value !== '.' ) ? 
        (keySymbol.includes(value))? last.ex : '' 
        : last.ex;

        return ({...last,ex:`${lastEx + value}`,respAct: false}); 
        
    });
};

//or clears the last digit entered depending on the status of the answer 
const handleReset = () => {
    setExercise(last =>{
        const last1 = (last.ex.length>1)? (last.ex.slice(0, -1)): '0';

        return (last.respAct)?
        ({ex: '0',res: 0, respAct: false}):
        ({...last,ex: last1});
    });
};

    return (
            <>
            <div className="col-sm-12">
                <div className="row">
                    {
                       keySymbol.map((simb,index) =>(
                        <div key={index.toString()} className="col-sm-3 contButton">
                            <button  type="button" className="btn" onClick={handleSymbolNumber} value={simb}>{simb}</button>
                        </div>
                       )) 
                    }
                </div>
            </div>
            <div className="col-sm-9">
                <div className="row">
                    {
                       keyNumeric.map((numb,index) =>(
                        <div key={index.toString()} className="col-sm-4 contButton">
                            <button type="button" className="btn" onClick={handleSymbolNumber}  value={numb}>{numb}</button>
                        </div>
                       )) 
                    }
                    <div className="col-sm-4 contButton">
                        <button type="button" className="btn btn-danger" onClick={handleReset}>AC</button>
                    </div>
                </div>
            </div>
            </>
    )
});
