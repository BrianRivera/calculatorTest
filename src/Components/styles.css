.calculator-container {
    margin-top: 15%;
}

.calculatorBase {
    border-radius: 5px;
    box-shadow: 0 5px 8px 0 rgba(0, 0, 0, 0.2), 0 9px 26px 0 rgba(0, 0, 0, 0.19);
}

.calculatorBase h3,
h5 {
    text-align: center;
    color: white;
}

.calculatorBase hr {
    background-color: white;
}

.resultCalculator {
    height: 150px;
    background-color: #333;
    border-radius: 10px 10px 0 0;
}

.numberCalculator {
    border-radius: 0 0 10px 10px;
    border-style: solid;
    border-width: 1px;
    border-color: #333;
}

.numberCalculator button {
    width: 90%;
    height: 90px;
    margin: 5px;
    font-size: 40px;
    box-shadow: 0 5px 8px 0 rgba(0, 0, 0, 0.2), 0 9px 26px 0 rgba(0, 0, 0, 0.19);
}

.contButton {
    text-align: center;
}

.resultDiv {
    text-align: center;
    margin-top: 50px;
}

.divBtnResult {
    box-shadow: 0 5px 8px 0 rgba(0, 0, 0, 0.2), 0 9px 26px 0 rgba(0, 0, 0, 0.19);
    width: 90%;
    height: 90%;
    font-size: 40px;
    border-radius: 10px;
}

.divBtnResult h1 {
    padding-block-start: 120%;
}

.resultDiv :hover {
    cursor: pointer;
    transition: background-color 0.5s ease;
    background-color: rgb(53, 14, 224) !important;
}