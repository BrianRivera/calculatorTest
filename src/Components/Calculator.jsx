
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import './styles.css';
import { Keys } from './Keys';
import { evaluate, round } from 'mathjs'

//if the initial value is not sent the initial value is equal to 0
export const Calculator = ( {initialValue = 0} ) => {

    /*
    hook useState to control both the result and the exercise
    */
    const [exercise, setExercise] = useState({
        ex: `${initialValue}`,
        res: initialValue,
        respAct: false
    });

    //function in charge of updating the result
    /*
    1.-handle errors using the try catch
    2.-generate the result using 'evaluate' from the mathjs library, 
       since according to developer.mozilla, the use of eval is not recommended
       https://developer.mozilla.org/es/docs/Web/JavaScript/Referencia/Objetos_globales/eval
    3.-update the result state 
    */

    const handleResult = () => {
     try {  
            const result =  round(evaluate(exercise.ex),4);
            setExercise({...exercise,res:result,respAct: true});
     } catch (error) {
            alert('Error, por favor revise el ejercicio');
     }
    }

    
    return (
        <div className="container calculator-container">
            <div className="col-md-6 calculatorBase mx-auto">
                <br/>
                
                <div className="resultCalculator">
                    <h5>{exercise.ex}</h5>
                        <hr/>
                    <h3>{exercise.res}</h3>
                </div>
                
                <div className="numberCalculator">
                    <div className="row">
                        <Keys setExercise={setExercise}></Keys>

                        <div className="col-sm-3 resultDiv">  
                            <div className="divBtnResult bg-primary text-white" onClick={handleResult}>
                                <h1>=</h1>
                            </div>
                        </div>

                    </div>
                </div>
                <br/>
            </div>
        </div>
    )
}


/*

return (
        <div className="container calculator-container">
            <div className="col-md-6 calculatorBase mx-auto">
                <br/>
                
                <div className="resultCalculator">
                    <h5>{exercise.ex}</h5>
                        <hr/>
                    <h3>{exercise.res}</h3>
                </div>
                
                <div className="numberCalculator">
                    <div className="row">
                        <Keys setExercise={setExercise}></Keys>

                        <div className="col-sm-3 resultDiv">  
                            <div className="divBtnResult bg-primary text-white" onClick={handleResult}>
                               <h1>=</h1>
                            </div>
                        </div>

                    </div>
                </div>
                <br/>
            </div>
        </div>
    )


*/

// propType, is used to designate the initialValue as a variable of type numeric
Calculator.propTypes = {
    initialValue : PropTypes.number
}

