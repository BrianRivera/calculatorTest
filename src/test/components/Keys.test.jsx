import '@testing-library/jest-dom';
import React from 'react';
import { shallow } from 'enzyme';
import { Keys } from '../../Components/Keys';




describe('Test in <Keys />', () => {

const setExercise = jest.fn();
const wrapper = shallow(<Keys setExercise={setExercise}></Keys>);

    test('component matches the snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('click on any button call the function ', () => {
        //simulates the click on button 1 sending the value of 1
        wrapper.find('button').at(0).simulate('click',{
            target:{
                value:'1'
            }
        });

        expect(setExercise).toHaveBeenCalled();
        //simulates the click of the 'AC' button
        wrapper.find('button').at(15).simulate('click');

        expect(setExercise).toHaveBeenCalledTimes(2);

    }); 
});
