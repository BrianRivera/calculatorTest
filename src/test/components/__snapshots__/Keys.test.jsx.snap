// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <Keys /> component matches the snapshot 1`] = `
<Fragment>
  <div
    className="col-sm-12"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-3 contButton"
        key="0"
      >
        <button
          className="btn"
          onClick={[Function]}
          type="button"
          value="+"
        >
          +
        </button>
      </div>
      <div
        className="col-sm-3 contButton"
        key="1"
      >
        <button
          className="btn"
          onClick={[Function]}
          type="button"
          value="-"
        >
          -
        </button>
      </div>
      <div
        className="col-sm-3 contButton"
        key="2"
      >
        <button
          className="btn"
          onClick={[Function]}
          type="button"
          value="*"
        >
          *
        </button>
      </div>
      <div
        className="col-sm-3 contButton"
        key="3"
      >
        <button
          className="btn"
          onClick={[Function]}
          type="button"
          value="/"
        >
          /
        </button>
      </div>
    </div>
  </div>
  <div
    className="col-sm-9"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-4 contButton"
        key="0"
      >
        <button
          className="btn"
          onClick={[Function]}
          type="button"
          value={1}
        >
          1
        </button>
      </div>
      <div
        className="col-sm-4 contButton"
        key="1"
      >
        <button
          className="btn"
          onClick={[Function]}
          type="button"
          value={2}
        >
          2
        </button>
      </div>
      <div
        className="col-sm-4 contButton"
        key="2"
      >
        <button
          className="btn"
          onClick={[Function]}
          type="button"
          value={3}
        >
          3
        </button>
      </div>
      <div
        className="col-sm-4 contButton"
        key="3"
      >
        <button
          className="btn"
          onClick={[Function]}
          type="button"
          value={4}
        >
          4
        </button>
      </div>
      <div
        className="col-sm-4 contButton"
        key="4"
      >
        <button
          className="btn"
          onClick={[Function]}
          type="button"
          value={5}
        >
          5
        </button>
      </div>
      <div
        className="col-sm-4 contButton"
        key="5"
      >
        <button
          className="btn"
          onClick={[Function]}
          type="button"
          value={6}
        >
          6
        </button>
      </div>
      <div
        className="col-sm-4 contButton"
        key="6"
      >
        <button
          className="btn"
          onClick={[Function]}
          type="button"
          value={7}
        >
          7
        </button>
      </div>
      <div
        className="col-sm-4 contButton"
        key="7"
      >
        <button
          className="btn"
          onClick={[Function]}
          type="button"
          value={8}
        >
          8
        </button>
      </div>
      <div
        className="col-sm-4 contButton"
        key="8"
      >
        <button
          className="btn"
          onClick={[Function]}
          type="button"
          value={9}
        >
          9
        </button>
      </div>
      <div
        className="col-sm-4 contButton"
        key="9"
      >
        <button
          className="btn"
          onClick={[Function]}
          type="button"
          value={0}
        >
          0
        </button>
      </div>
      <div
        className="col-sm-4 contButton"
        key="10"
      >
        <button
          className="btn"
          onClick={[Function]}
          type="button"
          value="."
        >
          .
        </button>
      </div>
      <div
        className="col-sm-4 contButton"
      >
        <button
          className="btn btn-danger"
          onClick={[Function]}
          type="button"
        >
          AC
        </button>
      </div>
    </div>
  </div>
</Fragment>
`;
